src/useImmediateUpdateEffect.ts

```diff
@@ -0,0 +1,31 @@
+import { DependencyList, useRef } from 'react'
+import useStableMemo from './useStableMemo'
+
+/**
+ * An _immediate_ effect that runs an effect callback when its dependency array
+ * changes. This is helpful for updates should must run during render, most
+ * commonly state derived from props; a more ergonomic version of https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops
+ *
+ * ```ts
+ * function Example({ value }) {
+ *   const [intermediaryValue, setValue] = useState(value);
+ *
+ *   useImmediateUpdateEffect(() => {
+ *     setValue(value)
+ *   }, [value])
+ * ```
+ */
+function useImmediateUpdateEffect(effect: () => void, deps: DependencyList) {
+  const firstRef = useRef(true)
+
+  useStableMemo(() => {
+    if (firstRef.current) {
```

Comment:
it seems a little unintuitive (or at least worthy of explanation) that this doesn't get run the first time

also – what's the concrete use case here? it seems like most of the time you'd just want to use `useMemo`

Reply:
The more specific hook here is `useStateFromProp`, That's the application we use this for mostly, when you need to sync a state value with the props. It's nicer to use a deps array vs usePrev and comparing, etc.

Reply:
err where is `useStateFromProp` defined?

Reply:
i meant conceptually, check out how we use this in bfly cloud for more specific context

Reply:
so maybe we should just expose the

```js
const [state, setState] = useStateFromProp(() => nextState, [prop])
```

?

this hook sort of feels weird to me, though, in that it's not expressing some sort of invariant or graph-y thing. like, we're sometimes setting the state, but it's not a consistent side effect... thing.

"state that resets whenever a prop changes" is... a fairly specific concept, maybe? i'm not even totally sure we're quite modelling it correctly – arguably the trigger there should be a navigation or something.

Reply:
I mean i use some version of this hook fairly often in places. Often to update state in response to some value change, but not always. Not sure what you mean by it being weird, is it b/c it doesn't run on the first call or b/c it fires immediately? Otherwise it's conceptually the same as the other effects. It'd be nicer if it just always ran but that doesn't compose well with other hooks.